Gh-380: Update user guide querying section (#393)
gchq/gaffer-doc: docs/user-guide/query/gaffer-syntax/filtering.md

# Filtering Data

> One of the key parts of querying data is being able to filter bits out you may

Reviewer: ```suggestion
One of the key parts of querying data is being able to filter out bits you may
```

not want. This page will cover some of the common ways you can filter the
returned data from an Operation to get the data you want.

In Gaffer the main way you filter data is by applying whats known as a `View` to

> a returned set of elements. A `View` as the name suggests, allows you to view

Reviewer: ```suggestion
a returned set of elements. A `View`, as the name suggests, allows you to view
```

> the data in a different way this can be either via a filter, aggregation,

Reviewer: ```suggestion
the data in a different way; this can be either via a filter, aggregation,
```

- **Transformation** - Transformations can be applied by providing Functions to
    transform properties and vertex values. This is a powerful feature, you can
    override the existing values or you can transform and save the new value
    into a new transient property.

- **Property Removal** - The relevant properties you want to be returned can be
    controlled. You can use either `properties` or `excludeProperties` to define
    the list of properties to be included or excluded.

## Filtering in Practice

We will demonstrate general filtering on a query. Take the following graph as a
basic example, where we have a `Person` node with ID `John` that has a few
`Created` edges with a `weight` property on them.

```mermaid
graph LR
    A(["Person

        ID: John"])
    --
    "Created
     weight: 0.2"
    -->
    B(["Software

        ID: 1"])
    A
    --
    "Created
     weight: 0.6"
    -->
    C(["Software

        ID: 2"])
```

Now say the `weight` property could represent how much a `Person` contributed to
creating something, and so we wanted to find only significant contributions. To
do this we can apply a filter to act as a threshold to only get edges with a

> `weight` more than a specific value.

Reviewer: The whole paragraph doesn't read too well. I would rephrase it to flow a little better. 
```suggestion

Lets say the `weight` property represents how much a `Person` contributed to
creating something and that we wanted to only find the most significant contributions. To
do this we can apply a filter to act as a threshold to only get edges with a
`weight` more than a specific value.
```

                                        or_equal_to = False
                                    )
                                )
                            ]
                        )
                    ]
                )
            )
        )
        ```

    === "Java"

        ```java
        // Define the View to use
        final View viewWithFilters = new View.Builder()
            .edge("Created", new ViewElementDefinition.Builder()
                    .preAggregationFilter(new ElementFilter.Builder()
                            .select("weight")
                            .execute(new IsMoreThan(0.4))
                            .build())
                    .build())
            .build();

        // Create the operation to execute
        final GetElements operation = new GetElements.Builder()
            .input(new EntitySeed("John"))
            .view(viewWithFilters)
            .build();

        graph.execute(operation, user);
        ```

To form relevant filters and queries its almost required that you are aware of
the graphs schema that is in use as this determines what properties and elements

> you can reference in your queries. For an introduction and background on

Reviewer: This sentence is a little confusing and could do with clarifying

Gaffer schemas [please see the guide](../../schema.md).

!!! tip
    As you can see filtering is based around predicates which are similar to if
    else statements in traditional programming. For a full list of available
    predicates refer to the [reference documentation](../../../reference/predicates-guide/predicates.md).

### Filtering Properties

If you are only interested in specific properties then it is more efficient to
tell Gaffer to only return those properties. This can be easily achieved by
applying the `properties` or `excludeProperties` field to a `View`.

> If we take a similar example as before but instead add a couple of properties

Reviewer: ```suggestion
If we take a similar example to before but instead add a couple of properties
```

to both the `Person` and `Created` elements to give a graph like the following:

```mermaid
graph LR
    A(["Person

        ID: John
        age: 34"])
    --
    "Created
     weight: 0.2
     hours: 100"
    -->
    B(["Software

        ID: 1"])
    A
    --
    "Created
     weight: 0.6
     hours: 800"
    -->
    C(["Software

        ID: 2"])
```

Now as before we can run a query on this graph to get the elements relevant to

> the `Person` node however, this time we will filter so that only specific

Reviewer: ```suggestion
the `Person` node. However, this time we will filter so that only specific
```

                        gaffer.ElementDefinition(
                            group = 'Created',
                            properties = [ "hours" ]
                        )
                    ]
                    entities = [
                        gaffer.ElementDefinition(
                            group = 'Person',
                            exclude_properties = [ "age" ]
                        )
                    ]
                )
            )
        )
        ```

    === "Java"

        ```java
        // Define the View to use
        final View viewWithFilters = new View.Builder()
            .edge("Created", new ViewElementDefinition.Builder()
                    .properties("hours")
                    .build())
            .entities("Person", new ViewElementDefinition.Builder()
                    .excludeProperties("age")
                    .build())
            .build();

        // Create the operation to execute
        final GetElements operation = new GetElements.Builder()
            .input(new EntitySeed("John"))
            .view(viewWithFilters)
            .build();

        graph.execute(operation, user);
        ```

## Transformation

> Its possible to apply a transformation to the output of a query this gives you

Reviewer: ```suggestion
It is possible to apply a transformation to the output of a query which then gives you the opportunity
```

an opportunity to manipulate the results into a more useful output.

> When a transform is applied the new results are save into whats known as a

Reviewer: ```suggestion
When a transform is applied the new results are saved into what is known as a
```

transient property. A transient property is just a property that is not

> persisted, simply created at query time by a transform function.

Reviewer: ```suggestion
persisted; simply created at query time by a transform function.
```

To use a transform you must use a transform function, this is a Java class
that extends the [`java.util.Function`](https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html)
class to take some input and give a new output. Commonly you would want to
write your own transform function class as it can be quite specific to your

> graph data and what analytics you want to get out of it. However, there is
> the [Koryphe module](../../../reference/functions-guide/koryphe-functions.md)
> that is included by default with Gaffer which provides a some default functions
> you can make use of.

Reviewer 2: ```suggestion
graph data and what analytics you want to get out of it. However, the
[Koryphe module](../../../reference/functions-guide/koryphe-functions.md)
(included by default with Gaffer) provides some default functions you can
make use of.
```

                            transient_properties = {'minutes': 'java.lang.Integer'},
                            transform_functions = [
                                gaffer.FunctionContext(
                                    selection = [ "hours" ],
                                    function = gaffer.MultiplyBy(by = 60),
                                    projection = [ "minutes" ]
                                )
                            ]
                        )
                    ]
                )
            )
        )
        ```

    === "Java"

        ```java
        final GetElements getEdgesWithMinutes = new GetElements.Builder()
            .input(new EntitySeed("John"))
            .view(new View.Builder()
                    .edge("Created", new ViewElementDefinition.Builder()
                            .transientProperty("minutes", Integer.class)
                            .transformer(new MultiplyBy(60))
                            .build())
                    .build())
            .build();

        graph.execute(getEdgesWithMinutes, user);
        ```

The `selection` in a transform is similar to the way we select properties and

> identifiers in a filter, and as demonstrated you can select (and also project)

Reviewer: ```suggestion
identifiers in a filter and as demonstrated you can select (and also project)
```

> any property but also or any of these unique identifiers:

Reviewer: ```suggestion
any property but also any of these unique identifiers:
```

- `VERTEX` - This is the vertex on an Entity.
- `SOURCE` - This is the source vertex on an Edge.
- `DESTINATION` - This is the destination vertex on an Edge.
- `DIRECTED` - This is the directed field on an Edge.
- `MATCHED_VERTEX` - This is the vertex that was matched in the query, either
  the `SOURCE` or the `DESTINATION`.
- `ADJACENT_MATCHED_VERTEX` - This is the adjacent vertex that was matched in
  the query, either the `SOURCE` or the `DESTINATION`. For example, if your seed

>   matches the source of the edge This would resolve to the `DESTINATION` value.

Reviewer: ```suggestion
  matches the source of the edge this would resolve to the `DESTINATION` value.
```

## Query-time Aggregation

Gaffer allows grouping results together based on their properties to form a new
result, this is known as aggregation. Aggregation can be applied at both data
ingest or query-time, this guide will focus on the latter but an overview of
both techniques is available in the [gaffer basics
guide](../../gaffer-basics/what-is-aggregation.md).

> Generally to apply aggregation at query-time you must override whats known as

Reviewer: ```suggestion
Generally to apply aggregation at query-time you must override what is known as
```
